feat: implement `cols_reorder()` for flexible column reordering

jrycw opened the pull request:
Hello team,  

This PR introduces `cols_reorder()`, allowing users to reorder columns in a specified order easily.  

While the existing `move_*()` functions are useful for repositioning a few columns, they don't provide fine-grained control over column order. I've often had to rely on `Pandas` or `Polars` for this task. The implementation of this function follows a straightforward approach, similar to the existing `move_*()` functions.

Feedback is always welcome!😊

jrycw committed:
Implement `cols_reorder()`

jrycw committed:
Implement `cols_reorder()`

jrycw committed:
Add tests for `cols_reorder()`

jrycw committed:
Add tests for `cols_reorder()`

jrycw committed:
Include `cols_reorder` in `_quarto.yml`

jrycw committed:
Include `cols_reorder` in `_quarto.yml`

zachvalenta committed:
doc: align getting started guide with readme

rich-iannone commented on great_tables/_spanners.py:
````diff
@@ -753,3 +753,58 @@ def cols_width(self: GTSelf, cases: dict[str, str] | None = None, **kwargs: str)
+    function: `cols_reorder()`. By passing a list of column names to the `columns=` parameter,
+    the table will be reordered accordingly:
+    ```{python}
+    col1, col2, col3, *cols = exibble.columns  # col1="num", col2="char", col3="fctr"
````
This is a really good example!

jrycw replied:
I've renamed the variables to use their actual column names instead of generic names like `col1`. This should improve readability.

rich-iannone commented on great_tables/_spanners.py:
```diff
@@ -753,3 +753,58 @@ def cols_width(self: GTSelf, cases: dict[str, str] | None = None, **kwargs: str)
+
+    if isinstance(columns, str):
+        columns = [columns]
+
```
I wonder if we should raise here, if only to provide a more user-friendly message. The user will instead see `"Reordering vars must contain all boxhead vars."` if not providing all columns, which is good for a developer-facing error message but it exposes some of the underlying implementation that the user might not know about.

jrycw replied:
Great suggestion! Should we raise the exception inside the function or modify the error message to be more user-friendly rather than developer-focused?

jrycw replied:
How about `"Column reordering must include all columns in the table."`?

rich-iannone replied:
I like it!

jrycw committed:
Make `Boxhead.order()` error message more user-friendly

jrycw committed:
Make `Boxhead.order()` error message more user-friendly

rich-iannone committed:
docs: add `style.css` entry to API reference docs (#624)

* Add docstring for CellStyleCss (style.css)

* Update _quarto.yml with style.css API entry

* Remove pip install of the pointblank package

* Include missing import in tab_spanner() example

jrycw committed:
Enhance the example for `cols_reorder()`

jrycw committed:
Enhance the example for `cols_reorder()`

machow committed:
feat: quarto option to disable table processing, warn on render (#611)

* feat: option to disable quarto table processing

* dev: upgrade ruff

* feat: run env specific checks at pre-render

* dev: revert ruff upgrade

* fix: ensure render checks get run on render

* fix: move pre render checks to where all paths call them

* feat: quarto render check handles percents

* fix!: enforce col widths have to be strings

* fix: use deprecation warning in cols_width instead

machow committed:
docs: mention .show() in a comment

rich-iannone committed:
fix: scientific notation zero padding when `exp_style=` used (#622)

* Use `rjust()` to correctly format sci notn w/ letter

* Add tests of fmt_scientific() method

* Fix incorrect testing of _format_number_compactly()

* Use pytest.mark.parametrize w/ fmt_scientific() tests

machow committed:
Merge pull request #627 from zachvalenta/align-get-starting-with-readme

doc: align getting started guide with readme

jrycw committed:
feat: add `cols_unhide()` method (#629)

* Implement `cols_unhide()`

* Add tests for `cols_unhide()`

* Include `cols_unhide()` in `_quarto.yml`

* Rename `Boxhead._set_cols_visible()` to `Boxhead._set_cols_info()` and update related logic

* Rename `Boxhead._set_cols_info()` to `Boxhead._set_cols_info_type()`

* Fix interlink for `cols_hide()` in `cols_unhide()` docs

tylerriccio33 committed:
Add inline_css argument (#633)

rich-iannone committed:
fix: update conditional statement in nanoplot code to guard against string values (single string or list containing any strings) (#623)

* Avoid use of _val_is_str() in conditional

* Add testcase for nanoplot with x and y values

* Reduce size of DataFrame for nanoplot test

* Refactor nanoplot test

rich-iannone committed:
fix!: move `css-inline` pkg to extra group (#634)

* Move css-inline pkg to extra group

* Check if css-inline is present in as_raw_html()

* Refactor and use correct module name

* Remove set_trace() statement

rich-iannone committed:
Add note about using .show("browser") when in VS Code

rich-iannone committed:
Modify wording slightly

rich-iannone committed:
Update docs/get-started/basic-styling.qmd

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>

machow committed:
Merge pull request #643 from posit-dev/docs-table-styling-show-browser

docs: add note about using `.show("browser")` when in VS Code

jrycw committed:
feat: Support pickling (#641)

* Support pickling for `GT`

* Move the `pickle` import statement inside `test_pickle()`

* Return `self._d` for `__getnewargs__` to support pickling

rich-iannone committed:
Remove unused sep_mark= from fmt_scientific() (#642)

rich-iannone committed:
feat: custom datetime formatting in `.fmt_datetime()` (#645)

* Add format_str arg to fmt_datetime()

* Add info on `format_str=` arg to docstring

* Implement use of `format_str=`

* Add test for .fmt_datetime(format_str=...)

* Refactor fmt_datetime_context()

* Add comments back

jrycw committed:
docs: suppress the warning message from `cols_width()` in the docs (#659)

* Correct the warning message in `_render_check_quarto()`

* Suppress the warning message from `cols_width()` in the docs

rich-iannone committed:
Update docstrings for date/time formatters (#658)

machow committed:
blog: polars .style post (#635)

* blog: polars .style post

* docs: change image links to jpg

* docs: update style blog post

* docs: full draft polars dot style

* docs: polars dot style tweak

* docs: more polars blog updates

* docs: polars style blog post

* docs: polars dot style quick fixes

Co-authored-by: Richard Iannone <riannone@me.com>

* docs: update polars dot style title

Co-authored-by: Richard Iannone <riannone@me.com>

* docs: polars dot style mention marco

* docs: add michael as author to blog post

---------

Co-authored-by: Richard Iannone <riannone@me.com>

dpprdan committed:
fix links (#670)

fixes two links

rich-iannone committed:
feat: add `compact=` arg to for `fmt_currency()` (#664)

* Add compact= parameter to fmt_currency()

* Add compact= arg to vals.fmt_currency()

* Add test of fmt_currency(compact=True)

* Refactor currency formatting based on code review

machow committed:
tests: add no pandas dep vals test (#689)

juleswg23 committed:
docs: add cols_label_rotate method (#696)

* First working version of cols_label_rotate

* docstrings partway done

* quarto change

* Test and snapshot

* docs update, TODO example text

* 8px padding included, and vertical-rl removed as it is identical to vertical-lr these cases

* Add to docstring

* update docs with one more example

* typos and cleanup

* stripped down the snapshot

* qmd doc formatting

* additional tests to pass codecov

* updating type literals and options on align

rich-iannone committed:
feat: add the `fmt_tf()` formatting method (#665)

* Add the fmt_tf() formatting method

* Remove unneeded check

* Remove comment

* Add example to `fmt_tf()` docs

* Refactor fmt_tf_context()

* Modify comments in fmt_tf_context()

* Add tests for fmt_tf()

* Move validation of tf_style= param

* Add several tests of fmt_tf()

* Add another test for fmt_tf()

* Add tests of _check_colors()

* Add test of pattern= use with fmt_tf()

* Add test where scalar colors= param used

* Refactor color application logic in fmt_tf_context()

* Add type annot to TF_FORMATS dict

* Reduce size of list in test

* Raise ValueError is input not bool/NA

* Add ValueError in else block for unexpected cases

* Put color-mapping logic in TfColorMap cls

* Better handle NA values (skip formatting if needed)

* Fix mistake in conditional logic (end with else)

* Refactor based on code review

* Remove reassignments; add check for na_val type

* Add warning about colors= unsupported in LaTeX

* Add a few more tests

rich-iannone committed:
docs: use improved interlinking available in quartodoc (#697)

* Increse min version requirement for quartodoc

* Enable autolink from quartodoc

* Use quartodoc autolinking syntax

* Add Quarto interlinks extension

* Add aliases in interlinks setup

* Use full path for method interlink

* Add updated interlinks extension to docs dir

* Update basic-column-labels.qmd

* Update basic-formatting.qmd

* Make corrections when linking vals functions

* Update basic-header.qmd

* Update basic-stub.qmd

* Use .qmd link to API reference

* Update basic-styling.qmd

* Remove interlink in header

* Update colorizing-with-data.qmd

* Update column-selection.qmd

* Update loc-selection.qmd

* Update nanoplots.qmd

* Update row-selection.qmd

* Update table-theme-options.qmd

* Update table-theme-premade.qmd

* Update targeted-styles.qmd

* Improve interlinking in GT class docstring

* Restore GT docstring to original form

We are going to treat docstrings differently w.r.t. interlinking so that IDEs don't show distracting characters like `~`

* Correct malformed interlink

* Add missing interlink in nanoplots page

* Use .qd-no-link in header interlink

rich-iannone committed:
Add underline to active link in top nav (#706)

rich-iannone committed:
Add the fmt_tf() method to the reference API (#704)

rich-iannone committed:
docs: improve docstring for the `as_raw_html()` method (#707)

* Improve docstring for as_raw_html()

* Tweak wording of final paragraph in description

* Remove redundant explanations of parameters

* Introduce param names in docstring description

machow committed:
feat: tab spanner delim (#647)

* Add scaffold for `tab_spanner_delim()`

* Simplify the spanner recording process

* Add `SpannerTransformer` implementation

* Add tests for `SpannerTransformer`

* Update `SpannerTransformer.get_rectangle()` implementation

* refactor: hold no column state in SpannerTransformer

* tests: flesh out tab_spanner_delim tests

* tests: add note about gather behavior

* docs: add tab_spanner_delim to api

* docs: add more tab_spanner_delim examples in docstring

* docs: tweak tab_spanner_delim docstring

* docs: shorten tab_spanner_delim intro

---------

Co-authored-by: jrycw <jerry@ycwu.space>

rich-iannone committed:
Introduce show() method in a callout (not in comment) (#712)

rich-iannone committed:
blog: Add post on marimo / GT reactive table functionality (#716)

* Add wip post on marimo and GT

* Delete great-tables-marimo-youtube.jpeg

* Add GIF

* Update post

* Add section on marimo's quarto plugin

* Rewrite intro to video paragraph

* Fix phrasing in final paragraph

* Remove mention of marimo islands

* Make changes based on code review suggestions

* Standardize spelling of 'marimo'

* Add short Wrapping Up section

* Make some final tweaks based on review

machow committed:
feat: remove pandas dependency in vals.fmt_integer (#719)

* feat: no pandas dependency in vals.fmt_integer

* refactor: move imports to top of module

---------

Co-authored-by: Richard Iannone <riannone@me.com>

juleswg23 committed:
Fix: `rescale_factor()` domain off by one (#718)

* correct off-by-one error

* snapshot update

rich-iannone committed:
fix: revise the internals of Google Font handling (#708)

* Add _to_html_style() method to GoogleFont cls

* Add methods for Google Font handling in GTData

* Add _google_font_imports attribute

* Use _add_google_font_import() in opt_table_font()

* Update tab_style() to use _add_google_font_import()

* Separate use of Google Fonts from additional CSS

* Get sorted list of Google font imports

* Add two tests of new Google Fonts implementation

* Simplify assert statements

* Use string instead of dict in _to_html_style()

* Add several tests of GoogleFont class

* Add tests of GoogleFont in CellStyleText

* Remove unused method and associated tests

* Remove unused _get_google_font_imports() method

* Remove unneeded tests regarding duplicate items

* Remove unneeded imports from test file

* Use frozenset for _google_font_imports; update method

* Remove underlying GFonts logic from _gt_data cls

juleswg23 committed:
fix: striping indexing issue (#701)

* fix striping indexing issue

* Snapshot test

---------

Co-authored-by: Richard Iannone <riannone@me.com>

ChristopherRussell committed:
Fix/spanner (#695)

* fix!: styling for higher-level spanners by adding id attributes

* tests: update render_html snapshot after spanner id fix

* Remove writing of spanner IDs and update snapshots

* Add test and snapshot

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>
Co-authored-by: Richard Iannone <riannone@me.com>

juleswg23 committed:
Docs blog images (#723)

* docs: outline blog post on images in tables

* dev: add svg.py to dev deps for blog post

* draft blog post

* add plotnine to dev-no-pandas

* update plotnine sizing [no ci]

* style and links

* typos, revisions

* further editing and update date

* modified order of the 3 methods (svg -> html -> plotnine)

* dev-related fixes

* docs: add frozen blog post

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>

mahdibaghbanzadeh committed:
Add truncate option to data_color (#673)

* added truncate argument to data_color

* test module for data_color truncate updated

* refactor: reduce number of logical branches

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>
Co-authored-by: Richard Iannone <riannone@me.com>

rich-iannone committed:
blog: add the `v0.18.0` intro post (#728)

* Add initial version of 0.18.0 intro post

* Update some explanations

* Update text explanation

* Correct typo in table title

* Modify blog post title

* Use autolinking for a few methods

* Revise list of features

* Autolink the remainder of the methods in the post

* Add leading dot before method names

FBruzzesi committed:
perf: Multiple nitpicks, some best practices and performances (#731)

* perf: Nitpicks in _gt_data.py

* _boxhead.py

* _export.py

* _formats.py

* _gt_data.py

* _helpers.py

* _locations.py

* _options.py

* _render.py

* _scss.py

* _spanners.py

* _styles.py

* _tbl_data.py

* _text.py

* _utils_nanoplots.py

* _utils_render_html.py

* _utils_render_latex.py

* _utils.py

* _data_color/

* fix issue in _utils_render_latex.py

* html improvements

* fix typo in nanoplots

* use tuple over set

* rstrip fix

* rstrip fix, pt2

* add _str_detect test with align right pattern

FBruzzesi committed:
fix: Various typos (#730)

* main codebase

* tests

* docs

FBruzzesi committed:
fix: Always access column names via `get_column_names` instead of `.columns` (#736)

* fix: use get_column_names in resolve_col_i

* use get_column_names in data_color

* add substitions test

FBruzzesi committed:
fix: Align pyarrow float64 type to the right to match pandas and polars behavior (#734)

* fix: Align pyarrow float64 type to the right to match pandas and polars behavior

* formatting

FBruzzesi committed:
test: refactor test_data_color_utils.py through parametrization

FBruzzesi committed:
parse version with _tbl_data _re_version func

FBruzzesi committed:
perf: Avoid double `clear` in polars (#729)

* perf: Avoid using polars map_elements

* linting

* rollback custom list parsing

tylerriccio33 committed:
Remove numpy from dependencies (#749)

* use narwhals to figure out if numpy is available

* use pure in linspace generation

* use random to generate jitter

* pure python normalize vals

* pure python scaling

* values to rgb overzealous infinity check on None value

* alter normalization tests

* drop numpy from main dependencies

* use bare import attempt instead of narwhals

* no filled just scaled

* move None check to top of rescale comp

* truncate logic on single statement

* add numpy integer abstract class

juleswg23 committed:
docs: Extending Great Tables (#743)

* first draft of get-started: Extending Great Tables

* Clarify example

* draft revisions based on comments

* typos, cleanup, links

* improving flow, typos

* pare down example

* Re-label to action verbs, and modify tense as needed

* typos, awkwardness

* phrasing

* add fmt_threshold example to "extending" section.

FBruzzesi committed:
Merge branch 'main' into test/refactor-data-color-utils

rich-iannone committed:
Modify FootnoteInfo class

rich-iannone committed:
Add the tab_footnote() method

rich-iannone committed:
Implement set_footnote for various location types

rich-iannone committed:
Remove unused tab_footnote() function and imports

rich-iannone committed:
Add footnote mark rendering to HTML table components

rich-iannone committed:
Improve footnote numbering to align with display order

rich-iannone committed:
Enable the footnotes_marks option in Options class

rich-iannone committed:
Enable the footnotes_marks parameter in tab_options()

rich-iannone committed:
Add support for custom footnote mark symbols

rich-iannone committed:
Add footnote marks to col labels in HTML rendering

rich-iannone committed:
Add several tests

rich-iannone committed:
Hide footnotes for hidden columns in HTML rendering

rich-iannone committed:
Add several tests

rich-iannone committed:
Update test_footnotes.py

rich-iannone committed:
Remove period from footnote marks in footer HTML

rich-iannone committed:
Add tab_footnote() to the API reference docs

rich-iannone committed:
Improve footnote handling and ordering in rendered tbl

rich-iannone committed:
Refactor footnote mark HTML generation

rich-iannone committed:
Align stub and data cell footnote ordering

rich-iannone committed:
Add snapshot test for footnotes in stub and body

rich-iannone committed:
Support unit notation in HTML/Markdown footnotes

rich-iannone committed:
Add unified HTML footer for source notes and footnotes

rich-iannone committed:
Update tests and snapshots

rich-iannone committed:
Modify docstring for tab_footnote()

rich-iannone committed:
Ensure that spanners and stubhead label can have footnotes

rich-iannone committed:
Update tests and snapshots

rich-iannone committed:
Remove unneeded utility function

rich-iannone committed:
Remove more unneeded code

rich-iannone committed:
Refactor footnote location hierarchy mapping

rich-iannone committed:
Incorporate footnote placement (auto/left/right)

rich-iannone committed:
Revise tests and snapshots

rich-iannone committed:
Refine numberlike detection

rich-iannone committed:
Update tests and snapshots

rich-iannone committed:
Add several tests for single-line source notes

rich-iannone committed:
Remove unneeded utility function

rich-iannone committed:
Add tests for some edge cases

rich-iannone committed:
Add tests of _get_column_index()

rich-iannone committed:
Add tests for _get_spanner_leftmost_column_index()

rich-iannone committed:
Refactor tests for tab_footnote()

rich-iannone committed:
Add opt_footnote_marks to API reference docs

juleswg23 committed:
Fix: allow seting `groupname_col` without `rowname_col` (#756)

* rename vars for clarity

* make group label rows even when the row_stub_column is not present

* add snapshot test

FBruzzesi committed:
Merge branch 'main' into test/refactor-data-color-utils

juleswg23 committed:
Fix `row_group_as_column = True` bug (#754)

* Add support for when has_two_col_stub is true

* ensure stubhead is aligned with rowname column - a little hacky maybe

* ensure is_row_group_cell is never null

* Include styles

* re-structure to be a rowspanning th element, rather than blank cells

* revert commit 20670df: ensure stubhead is aligned with rowname column

* refactor to add the row cell in the same place that group rows are added

* remove todo

* rename vars for clarity

* make group label rows even when the row_stub_column is not present

* add snapshot test

* update old snapshot test and remove comment

* add tests

* remove shows from tests

* complete merge from main

* ruff whitespace

juleswg23 committed:
Fix: `_ideal_fgnd_color()` takes alpha into account (#747)

* add _alpha_composite_with_white()

* add test for new function

* _remove_alpha is now unused code

* note todo rgb_to_hex

juleswg23 committed:
Fix: Row striping font color contrast with background (#745)

* Include row_striping_background_color and font_color in compiled scss

* remove duplicate font_color_vars

* Fix default value for row_striping_background_color OptionsInfo

* regenerate snapshots

rich-iannone committed:
Refactor footnote location handling to use Loc objects

rich-iannone committed:
Refactor footnote application logic in HTML rendering

rich-iannone committed:
Split long HTML string in two for better readability

rich-iannone committed:
Merge branch 'main' into feat-tab-footnote

rich-iannone committed:
Refactor footnote handling to use FootnoteEntry

rich-iannone committed:
Remove unneeded text processing for footnotes

rich-iannone committed:
Refactor footnote mark HTML creation

rich-iannone committed:
Fix footnote mark type handling

rich-iannone committed:
Update _locations.py

rich-iannone committed:
Throw error if writing a LaTeX tbl with footnotes

rich-iannone committed:
Update as_latex() docstring with footnotes limitation

rich-iannone committed:
Add test to check for as_latex() raising w/ footnote

rich-iannone committed:
docs: add notice of 2025 Posit Table Contest in the README (#776)

* Add notice of 2025 Table Contest in the README

* Add notice to landing page of project website

rich-iannone committed:
Include Posit badge in header (#777)

juleswg23 committed:
Feat: add `grand_summary_rows()` method (#765)

* a very very rough draft, does not work for polars yet

* set stub and boxhead to include all rows in final table

* add create_no_row_frame and refactor insert_row

* add SummaryRows to GTData object

* refactor modify rows

* add merge_summary_rows to append to body

* replace len with n_rows

* update build step

* add comments mapping out rendering plan

* new locations for grand summary (and summary)

* condensing locations

* add getter for summary rows

* begin design on rendering top grand summary rows

* refactor _create_row_component_h(), outside of create_body_component_h(), called also for making summary rows

* remove merge_summary_rows

* remove commented out code

* Change SummaryRowInfo values from list to dict

* reorder functions

* remove dead code

* support grand summary rows at bottom of table

* location mask class var added

* support sum

* support opt_stylize and other tab_options

* Apply special classes to get double border

* clean up add summary and set up styling

* rename for clairity

* experimenting with styles on grandSummaryStub

* remove prints

* Add stub column when none exists for summary rows (this approach does not involve boxhead)

* target location for LocGrandSummaryStub

* style LocGrandSummaryStub and LocGrandSummary

* Handle special cases for summary rows with group stub columns

* locations docstrings

* adding new locs to quarto

* more documentaiton

* accept summaryFn and label

* possible approach to fmt in grand Summary Rows, WIP

* refactor summary rows to mapping, and add summary rows grand attribute to gt_data

* use eval_aggregate for summary rows

* remove dead code

* docstring

* testing docstring in site preview

* fix build

* refactor to rely on class attribute to determine if grand summary

* test eval_aggregate

* grand summary rows tests added

* fix kitchen sink example

* ensure example compiles

* locations tests

* snapshot updates, cover case in utils_render_html

* main docstring

* get started documentation

* docs nitpicks

* remove unused attribute

* refactor: clean up some mutation, make kw-only

* fix: areas in grand summary code relying on mutation

* refactor: replace for loop with gen expr

* chore: move import under TYPE_CHECKING

* refactor: add summary_placeholder stub column type

* tests: remove unused output from assertion

* fix: allow summarizing to non-int, float

* docs: start some technical notes

* ci: give quarto action our gh token

* docs: fix now mandatory kw arg

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>

rich-iannone committed:
Add LaTeX support for the stub and row groups

rich-iannone committed:
Add LaTeX rendering tests for stubs and row groups

rich-iannone committed:
Remove2025 Table Contest submission info notice (#789)

machow committed:
feat: support polars expressions in vals functions (#793)

* feat: support polars expressions in vals functions

* tests: check val.fmt_*() function on expression

* feat: better type overloads for val expression handling

* fix: add wraps to our decorator

* fix: support py3.9 typing

FBruzzesi committed:
merge main, solve conflicts

jiboncom committed:
Add Markdown/HTML to Latex support

jiboncom committed:
Add tests for Markdown/HTML to LaTeX conversion

jiboncom committed:
Fixing code formatting

rich-iannone committed:
Update copyright end year to 2026 (#804)

kianmeng committed:
Fix typos (#798)

Found via `codespell -S *.csv -L ser,te,revered,indx`

rich-iannone committed:
Merge main into feat-tab-footnote

rich-iannone committed:
Normalize whitespace in test_export.ambr snapshots

rich-iannone committed:
Update test_export snapshots after merge

rich-iannone committed:
Add Plausible analytics to project website (#806)

rich-iannone committed:
feat: add `fmt_engineering()` (#786)

* Add engineering notation formatting method

* Update _formats_vals.py

* Update vals.py

* Update test_formats.py

* Add GT.fmt_engineering to reference API

* Add tests for some fmt_engineering() edge cases

* Add tests for engineering notn formatter function

* Refactor vals.py imports and update Ruff config

* Remove redundant import in fmt_engineering_context()

* Remove unused engineering notation function

* Remove sep_mark parameter from engineering format functions

* Reorder and clean up imports in _formats_vals.py

* Simplify engineering format test cases

* Remove sep_mark arg from eng formatters

* Include simple example to start off section

machow committed:
refactor: clean up auto-align code (#810)

* chore: update snapshots for pandas/polars dtype changes

Pandas 2.x uses "str" dtype for string columns instead of "object",
and polars list columns now get "center" alignment. These snapshot
updates reflect the current behavior with the existing codebase.

* Revert "chore: update snapshots for pandas/polars dtype changes"

This reverts commit 7764660cb652442dca54c9559a616286c6b4add0.

* fix: support pandas 3.x "str" dtype in auto-alignment

Pandas 3.x uses "str" dtype for string columns instead of "object".
This change:
- Adds "str" to the set of dtypes that get left-aligned
- Extends number-like detection to also check "str" dtype columns,
  so string columns with numeric-looking values still get right-aligned

This maintains consistent behavior across pandas 2.x and 3.x.

* refactor: extract auto-alignment logic to _tbl_data_align module

Extract dtype classification logic from Boxhead.align_from_data() into
a dedicated module using singledispatch for cleaner separation of concerns.

Changes:
- Create _tbl_data_align.py with classify_dtype_for_alignment() and
  is_number_like_column() functions
- Simplify align_from_data() to use the new module
- Add comprehensive tests in test_tbl_data_align.py (32 tests)

The refactor preserves existing behavior:
- Uses re.match (not re.search) to match original _str_detect behavior
- Supports both pandas 2.x ("object") and 3.x ("str") string dtypes
- Number-like string detection works for both dtype variants

* test: add PyArrow tests for dtype classification

Add tests for PyArrow backend coverage in test_tbl_data_align.py:
- TestClassifyDtypePyArrow class with numeric, string, bool, date tests
- PyArrow number-like detection tests
- PyArrow integration tests with GT

Coverage for _tbl_data_align.py increased from 74% to 87%.

rich-iannone committed:
Add tests for the opt_all_caps() method (#812)

jrycw committed:
Expose val_fmt_engineering (#808)

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>
Co-authored-by: Richard Iannone <riannone@me.com>

rich-iannone committed:
feat: add the `.opt_css()` method (#775)

* Add the opt_css() method

* Add opt_css() to API reference

* Add several tests of opt_css()

* Fix comment

* Filter empty strings from table_additional_css option

* Refactor and expand opt_css test coverage

* Clarify opt_css() usage in docstring

* Clarify note in opt_css() docstring

rich-iannone committed:
feat: add the `.cols_merge()` method (#780)

* Add column merging through cols_merge() method

* Add several tests for cols_merge()

* Improve cols_merge() pattern handling

* Extract column-merge logic to _cols_merge.py

* Extract ColMergeInfo and refactor column merge

* Add tests for ColMergeInfo and merge_pattern

* Switch cols_merge pattern indexing to 0-based

* Add GT.cols_merge to quartodoc list

* Remove unused _extract_pattern_columns helper

* Refactor cols merge: pattern and NA handling

* Refactor ColMergeInfo tests to use merge()

* Fix example indices in cols_merge() docstring

* Move column-merge pattern utils to _cols_merge.py

thriller08 committed:
feat: add container padding to tab_options (#801) (#802)

* feat: add container padding to tab_options (#801)

This allows the user to provide container vertical and horizontal
padding values to the tab_options function for more precise control
over layout when rendered to HTML.

* Reorder container args and tidy docs

---------

Co-authored-by: Richard Iannone <riannone@me.com>

jrycw committed:
Improve the docstring for `from_column()` (#766)

Co-authored-by: Richard Iannone <riannone@me.com>

tylerriccio33 committed:
add int to type signature (#800)

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>

rich-iannone committed:
Merge branch 'main' into test/refactor-data-color-utils

juleswg23 committed:
Docs: examples page -> gt-extras example (#759)

* first pass, adding gt-extras example

* Clean up page as a whole

* lock gt-extras at current version, 0.0.7 (or newer)

* no release version

* add fallback version

* revert last two commits, instead resolve #760

* install gt-extras "manually" in CI docs workflow

* fetch depth: 0

* change with: fetch-depth: 0 in all places

* add gt-extras back as a dependency

* delete pip install gt-extras

* docs: update examples to work with pandas v3

* docs: include gini example in subdir now

---------

Co-authored-by: Richard Iannone <riannone@me.com>
Co-authored-by: Michael Chow <mc_al_github@fastmail.com>

schmidma committed:
fix: update Polars selector type annotation to use `Selector` (#770)

* fix: update Polars selector type annotation to use `Selector`

Polars replaced `_selector_proxy_` with a public `Selector` class
(https://github.com/pola-rs/polars/pull/23351). This caused type
checking imports to fail. This updates the type annotation import and
alias accordingly.

* fix: use PlSelector consistently as type

* fix: _databackend can decide to not fail for non-existing imports

* chore: remove dead code

* fix: use explicit ImportError in except block

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>

FBruzzesi committed:
fix(typing): Annotate `pipe` first argument type as `GT` (#757)

* fix(typing): pipe missing first argument type

* import directly

---------

Co-authored-by: Richard Iannone <riannone@me.com>

juleswg23 committed:
CI: Fix checkout version tag (#768)

* add fetch depth

* add gt-extras to dependencies

* ci: use fetch-tags so setuptools_scm can version during jobs

* revert: jules was right the whole time

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>
Co-authored-by: Richard Iannone <riannone@me.com>

rich-iannone committed:
Merge branch 'main' into feat-latex-output-enhance

machow committed:
feat: add cols_label_with() method for function-based column relabeling (#816)

* feat: add cols_label_with() method for function-based column relabeling

Simplified version of PR #626. Adds cols_label_with(columns, fn) that
applies a callable to column labels. Columns-first arg order matches
cols_*() convention. No unit syntax handling, no Polars expression
support (tracked in #648).

* docs: clarify cols_label_with fn param operates on column names, not display labels

---------

Co-authored-by: Richard Iannone <riannone@me.com>

rich-iannone committed:
docs: correct docstring for the `fns=` parameter in the `fmt()` method (#818)

* Restrict fmt() fns= param to single FormatFn

* Clarify fns= param in fmt() docstring

* Validate and wrap fns argument in fmt()

* Handle FormatFns and callable fns in fmt

juleswg23 committed:
Fix: Modify the group labels when render_formats() is called (#769)

* modify the stub formats when render_formats() is called

* Refactor rowgroup label update logic into helper fn

* Add tests for row group label formatting

* Move update_group_row_labels into Stub

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>
Co-authored-by: Richard Iannone <riannone@me.com>

jrycw committed:
docs: add an example demonstrating `define_units()` usage (#446)

* Add an example demonstrating `define_units()` usage

* Add `GT.cols_label()`

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>
Co-authored-by: Richard Iannone <riannone@me.com>

rich-iannone committed:
Merge branch 'main' into test/refactor-data-color-utils

rich-iannone committed:
Merge branch 'main' into main

jrycw committed:
Refactor `GT.opt_all_caps()` (#436)

* Refactor `GT.opt_all_caps()`

* Update flawed logic

* Modify the `locations` parameter of `opt_all_caps()` to accept `Loc` objects

* Fix row_groups typos and import ordering

* Update tests to use loc.row_groups

* Make opt_all_caps() locations= default to None

* Uncomment and enable opt_all_caps() test assertions

* Deprecate string locations in opt_all_caps()

* Validate string locations in opt_all_caps()

* Warn before validating string locations

* Refine opt_all_caps location type hints

---------

Co-authored-by: Richard Iannone <riannone@me.com>

rich-iannone committed:
Merge branch 'main' into test/refactor-data-color-utils

lorenzo-w committed:
Fix #791 by handling `interactive_data_values` properly (#792)

* Fix #791 on upstream repo

* Refactor pipe function to use Concatenate in type hints

Updated the pipe function signature to use Concatenate for better type hinting.

* Add tests for nanoplot interactive option

* tests: test interactive_data_values svg

---------

Co-authored-by: Richard Iannone <riannone@me.com>
Co-authored-by: Michael Chow <mc_al_github@fastmail.com>

tylerriccio33 committed:
Pre commit makefile targets (#767)

* do not track lockfile

* add targets

---------

Co-authored-by: Richard Iannone <riannone@me.com>

KristijanArmeni committed:
fix: Convert Polars container dtype entries to list in _get_cell (#778) (#794)

* fix: convert container type entries to list in polars tables

* refactor: use tuple for second arg in isinstance

* test: test _get_cell for container dtypes

* Guard None before to_list() for list cells

* Handle None in container dtype tests

* Include pl.Array in list column detection

* refactor: simplify list-column check

---------

Co-authored-by: Michael Chow <mc_al_github@fastmail.com>
Co-authored-by: Richard Iannone <riannone@me.com>

FBruzzesi committed:
fix: Address slicing for pyarrow array in `data_color` (#741)

* fix: Address slicing pyarrow array in data_color

* rename function

* collapse pyarrow into one func:

---------

Co-authored-by: Richard Iannone <riannone@me.com>
Co-authored-by: Michael Chow <mc_al_github@fastmail.com>

machow committed:
docs: septa timetable blog post (#636)

* docs: septa timetable blog post

* docs: update septa blog post

* blog: update septa blog post

* docs: update again

* docs: add metadata to septa blog

jrycw committed:
Fix formatting issue in `fmt_threshold()` function (#821)

Fix #797

Co-authored-by: Richard Iannone <riannone@me.com>

rich-iannone committed:
Update .gitignore

rich-iannone committed:
Update CODEOWNERS

rich-iannone committed:
fix: update snapshots affected by updated `css_inline` library (#827)

* Dedent closing triple quotes in gt.py

* Trim whitespace and fix snapshot name

* Update test_export snapshot spacing

* Pin css-inline to >=0.20.2,<0.21

rich-iannone committed:
Update CODEOWNERS

mrapacz committed:
fix: prepend charset meta tag to HTML in save() (#823)

minimav committed:
Issue 283 nano plot scaling (#404)

* Filter rows correctly for nanoplot scaling

* Tests row selection used for scaling nanoplots

* Add comments to nanoplot SVG comparison tests

---------

Co-authored-by: Richard Iannone <riannone@me.com>

rich-iannone committed:
Merge main into feat-tab-footnote

rich-iannone committed:
Drop locnum from FootnoteInfo and adjust mappings

rich-iannone committed:
Trim excessive trailing whitespace in snapshots

rich-iannone committed:
Refactor and expand footnotes tests

rich-iannone committed:
Apply footnotes to group labels & grand summaries

rich-iannone committed:
Add tests for tab_footnote() footnote marks

rich-iannone committed:
Merge pull request #763 from posit-dev/feat-tab-footnote

rich-iannone committed:
Merge branch 'main' into main

rich-iannone committed:
Merge branch 'main' into test/refactor-data-color-utils

rich-iannone committed:
Clean up some data_color() test cases

rich-iannone committed:
Merge pull request #740 from FBruzzesi/test/refactor-data-color-utils

test: refactor `test_data_color_utils.py` through parametrization

rich-iannone committed:
Merge remote-tracking branch 'origin/main' into feat-latex-output-enhance

# Conflicts:
#	great_tables/_utils_render_latex.py
#	tests/test_utils_render_latex.py

rich-iannone committed:
Merge pull request #784 from posit-dev/feat-latex-output-enhance

feat: Enhance LaTeX output tables by adding support for the stub and row groups

rich-iannone committed:
Use multimark; remove markdown packages

rich-iannone committed:
Use multimark for Markdown rendering

rich-iannone committed:
Adjust HTML->LaTeX expectations; add Markdown tests

rich-iannone committed:
Merge branch 'main' into main

rich-iannone committed:
Merge pull request #803 from jiboncom/main

Add Markdown & HTML to LaTeX conversion

rich-iannone committed:
Merge branch 'cols-reorder' of https://github.com/jrycw/great-tables into pr/625

# Conflicts:
#	great_tables/_spanners.py
#	great_tables/gt.py
#	tests/test_spanners.py

rich-iannone committed:
Improve cols_reorder() docstring and examples

rich-iannone commented:
Sorry this took so very long to review, but going to merge now!

jrycw commented:
Thanks @rich-iannone. The docstring and examples are much cleaner than my original draft. Great work.
